pcsx2/DEV9/AdapterUtils.cpp

```diff
@@ -244,30 +246,91 @@ std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 
 	return std::nullopt;
 }
-#elif defined(__POSIX__)
-#ifdef __linux__
+#else
 std::optional<MAC_Address> AdapterUtils::GetAdapterMAC(Adapter* adapter)
 {
-	struct ifreq ifr;
-	strcpy(ifr.ifr_name, adapter->ifa_name);
+	MAC_Address macAddr;
+
+/*
+ * BSD 4.4 defines the size of an ifreq to be
+ * max(sizeof(ifreq), sizeof(ifreq.ifr_name)+ifreq.ifr_addr.sa_len
+ * However, under earlier systems, sa_len isn't present, so the size is 
+ * just sizeof(struct ifreq)
+ */
+#ifdef HAVE_SA_LEN
+#ifndef max
+#define max(a, b) ((a) > (b) ? (a) : (b))
+#endif
+#define ifreq_size(i) max(sizeof(struct ifreq), \
+	sizeof((i).ifr_name) + (i).ifr_addr.sa_len)
+#else
+#define ifreq_size(i) sizeof(struct ifreq)
+#endif /* HAVE_SA_LEN*/
+
+	int sd = socket(AF_INET, SOCK_DGRAM, IPPROTO_IP);
+	if (sd < 0)
+	{
+		return std::nullopt;
+	}
 
-	int fd = socket(AF_INET, SOCK_DGRAM, 0);
-	int ret = ioctl(fd, SIOCGIFHWADDR, &ifr);
-	close(fd);
+	char buf[1024] = {0};
+	struct ifconf ifc;
+	memset(buf, 0, sizeof(buf));
+	ifc.ifc_len = sizeof(buf);
+	ifc.ifc_buf = buf;
 
-	if (ret == 0)
-		return *(MAC_Address*)ifr.ifr_hwaddr.sa_data;
+	if (ioctl(sd, SIOCGIFCONF, (char*)&ifc) < 0)
+	{
+		close(sd);
+		return std::nullopt;
+	}
 
-	return std::nullopt;
-}
+	for (int i = 0; i < ifc.ifc_len; i += ifreq_size(*ifrp))
```

Comment:
The loop appears to only be necessary for the `AF_LINK` case, since you can directly copy `adapter->ifa_name` into `ifr.ifr_name` for the `SIOCGIFHWADDR`/`SIOCGENADDR` `ioctl` calls.

Reply:
🆗